rspace/src/main/scala/coop/rchain/rspace/HotStore.scala

```diff
@@ -32,261 +34,388 @@ trait HotStore[F[_], C, P, A, K] {
   def snapshot(): F[Snapshot[C, P, A, K]]
 }
 
-final case class Cache[C, P, A, K](
-    continuations: TrieMap[Seq[C], Seq[WaitingContinuation[P, K]]] =
-      TrieMap.empty[Seq[C], Seq[WaitingContinuation[P, K]]],
-    installedContinuations: TrieMap[Seq[C], WaitingContinuation[P, K]] =
-      TrieMap.empty[Seq[C], WaitingContinuation[P, K]],
-    data: TrieMap[C, Seq[Datum[A]]] = TrieMap.empty[C, Seq[Datum[A]]],
-    joins: TrieMap[C, Seq[Seq[C]]] = TrieMap.empty[C, Seq[Seq[C]]],
-    installedJoins: TrieMap[C, Seq[Seq[C]]] = TrieMap.empty[C, Seq[Seq[C]]]
+final case class HotStoreState[C, P, A, K](
+    continuations: Map[Seq[C], Seq[WaitingContinuation[P, K]]] =
```

Comment:
Good question!

Difference:

1. TrieMap is a trie-based **concurrent** scalable map implementation. which is like the `history trie` you are trying to change.It provides an **efficient, non-blocking, O(1) time snapshot operation** while the performance is a little bit slower than `Map` which is just a java `ConcurrentHashMap`.(Getit  from some reading)
2. `TrieMap` is a **mutable** data structure while `Map` here is an **immutable** data structure. Functional programming languages favor immutable data structures which can provide safe concurrent operations.

> which benefit of this change?

There is a bug on hotstore which the data is not consistent after the operation. However, I can not really know the deep detail of the bug which you have to replay thoudsands of time to reproduce it. @nzpr got a pr the change the data into immutable ones and I just rebase the work from him.

Reply:
Thanks, for the answers. I have the same structure (TrieMap) for hashing in HistoryStore. And in my code, I don't use the snapshot function. I don't know, what better use in my case. My priority is speed. @tgrospic  @nzpr @zsluedem what do you think?

Reply:
Oh. In my opinion, You don't have to compare your implementation with scala TrieMap. They are for different use cases. 
1. Your implementation should be able to used either in-memory storage or disk storage while scala TrieMap is only memory. 
2. Your implementation should be able to keep the history track of storage (by stateHash) while scala TrieMap don't do that(But I think you can consider snapshot as somthing similar.)

Reply:
We want to allow HotStore state to change only atomically. TrieMap allows concurrent access which has been used previously. New implementation puts the whole `HotStoreState ` inside a `Ref`, which manages concurrent access. In general, we don't want to have TrieMap-s, because it allows concurrency that is detached from our scheduler and managed by JVM. We want to use `Concurrent` typeclass for our F[_] (which is monix.Task) in places that require concurrency.